app/src/main/java/chat/rocket/android/activity/ServerConfigActivity.java

```diff
@@ -77,14 +83,27 @@ protected void onPause() {
   }
 
   private void onRenderServerConfigError(ServerConfig config) {
-    if (config != null) {
+    if (config.getState() == ServerConfig.STATE_CONNECTION_ERROR) {
       sessionObserver.unsub();
       showFragment(new RetryConnectFragment());
     } else {
       sessionObserver.sub();
     }
   }
 
+  private void continueWithServerConfigSession(final Session session) {
+    fetchPublicSettings()
+        .continueWith(task -> {
+          registerForPush();
+          return task;
+        })
+        .continueWith(task -> {
+          onRenderServerConfigSession(session);
+          return task;
+        })
+        .continueWith(new LogcatIfError());
+  }
```

Comment:
As continueWithServerConfigSession and `SessionObserver#onLogin` uses the same observer implementation, they will be called almost the same moments.

However not guaranteed to be called before user login (even if the logic is put in the Activity)
It depends on the implementation of the server.

SessionObserver just observes whether verified auth token exists.
CurrentUserObserver just observes whether an user with email address exists.

```
Text	Thu Dec 08 20:05:25 JST 2016	Thu Dec 08 20:05:25 JST 2016	a["{\"msg\":\"added\",\"collection\":\"users\",\"id\":\"wffqrSXNojgMnuz8v\",\"fields\":{\"emails\":[{\"address\":\"yusuke.iwaki@example.com\",\"verified\":true}],\"username\":\"YusukeIwaki\"}}"]

Thu Dec 08 20:05:25 JST 2016	Thu Dec 08 20:05:25 JST 2016	a["{\"msg\":\"result\",\"id\":\"1\",\"result\":{\"id\":\"wffqrSXNojgMnuz8v\",\"token\":\"0JdBqCN90XpH・・・・・・・dB_Ti7L3bB0\",\"tokenExpires\":{\"$date\":1488897869020}}}"]
```

When I observed DDP messages last month, it seems session observers are called AFTER user logged in.